app/controllers/concerns/saml_idp_logout_concern.rb

```diff
@@ -85,7 +85,7 @@ def prepare_saml_logout_response
   end
 
   def prepare_saml_logout_request
-    validate_saml_request
+    return if validate_saml_request
```

Comment:
The bug that we used to have was that we were expecting the SAML request validation to stop the logout process from continuing if the request was invalid, but because we are calling `validate_saml_request` from within a method inside the `logout` action, it will still continue executing the rest of the method if the request is invalid.

While your solution works, it is hard to understand IMO. When I read `return if validate_saml_request`, at first glance it means to me that if the request was valid, then we want to stop preparing the logout request, which is the opposite of what is actually happening.

I propose an alternative solution, which is to add a `before_action` that validates the request, like so:

```ruby
before_action :validate_saml_logout_request, only: :logout

def validate_saml_logout_request(raw_saml_request = params[:SAMLRequest])
  track_logout_event
  return unless raw_saml_request

  decode_request(raw_saml_request)
  head :bad_request unless valid_saml_request?
end
```

and then remove the call to `track_logout_event` from the top of the `logout` action.

I would also suggest updating `track_logout_event` to capture whether or not the SAML request was valid:
```ruby
def track_logout_event
  saml_request = params[:SAMLRequest]
  result = {
    sp_initiated: saml_request.present?,
    oidc: false,
  }
  result.merge!(saml_request_valid: valid_saml_request?) if saml_request
  
  analytics.track_event(Analytics::LOGOUT_INITIATED, result)
end
```
Ideally, we would also update the `saml_idp` gem so that the validation returns a rich object that we can fetch errors from so we can see exactly why the request failed validation. Currently, the gem just outputs the failure reason to the Rails log.

Also, note that I changed the error from a 403 forbidden to a 401 bad request, which I think is more appropriate in this case.

Thoughts?

Reply:
Great ideas. The gem's method name validate_saml_request definitely gives no indication that nil/false is the good condition so the name is misleading.  Their 403 error is also not right but I think you mean we should return 400 bad request (not 401).  I'd prefer their gem did the work and fix these issues but this is a good interim solution (we are effectively fixing their method and putting it in our code).  I was planning on doing a separate pr for more descriptive errors like 'Signature is invalid' (LG-224).  A 400 bad request still doesn't give much but at least it's an improvement.